File: backend/app/models/promotions.py
"""Promotional Schemes Auto-Apply models."""
import uuid
import enum
from sqlalchemy import (
    Column, String, Text, Numeric, Boolean, Integer,
    ForeignKey, Enum, Date, DateTime, UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.db.base import Base, TimestampMixin


# ── Enums ──────────────────────────────────────────────────────────────────────

class SchemeStatus(str, enum.Enum):
    DRAFT     = "DRAFT"
    APPROVED  = "APPROVED"
    ACTIVE    = "ACTIVE"
    EXPIRED   = "EXPIRED"
    SUSPENDED = "SUSPENDED"
    ARCHIVED  = "ARCHIVED"


class SchemeType(str, enum.Enum):
    BUY_X_GET_Y           = "BUY_X_GET_Y"
    PERCENT_DISCOUNT      = "PERCENT_DISCOUNT"
    FIXED_DISCOUNT        = "FIXED_DISCOUNT"
    TIERED_DISCOUNT       = "TIERED_DISCOUNT"
    QTY_BREAK_PRICE       = "QTY_BREAK_PRICE"
    SPEND_BASED           = "SPEND_BASED"
    MIX_AND_MATCH         = "MIX_AND_MATCH"
    BUNDLE                = "BUNDLE"
    FREE_GOODS_DIFF_SKU   = "FREE_GOODS_DIFF_SKU"
    CHANNEL_DEAL          = "CHANNEL_DEAL"


class TriggerBasis(str, enum.Enum):
    SKU         = "SKU"
    CATEGORY    = "CATEGORY"
    BRAND       = "BRAND"
    ORDER_TOTAL = "ORDER_TOTAL"
    QUANTITY    = "QUANTITY"
    VALUE       = "VALUE"
    MIX_SET     = "MIX_SET"


class RewardType(str, enum.Enum):
    FREE_GOODS      = "FREE_GOODS"
    PERCENT_DISCOUNT = "PERCENT_DISCOUNT"
    FIXED_DISCOUNT  = "FIXED_DISCOUNT"
    SPECIAL_PRICE   = "SPECIAL_PRICE"
    BUNDLE_PRICE    = "BUNDLE_PRICE"


class PromoApplicationType(str, enum.Enum):
    AUTO             = "AUTO"
    MANUAL_APPROVED  = "MANUAL_APPROVED"
    MANUAL_OVERRIDE  = "MANUAL_OVERRIDE"
    REJECTED         = "REJECTED"


class PromoImpactType(str, enum.Enum):
    DISCOUNT         = "DISCOUNT"
    FREE_GOODS       = "FREE_GOODS"
    SPECIAL_PRICE    = "SPECIAL_PRICE"
    ORDER_DISCOUNT   = "ORDER_DISCOUNT"
    BUNDLE_ADJUSTMENT = "BUNDLE_ADJUSTMENT"


class OverrideStatus(str, enum.Enum):
    PENDING  = "PENDING"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"


class PromoAIAgentType(str, enum.Enum):
    CONFLICT_ADVISOR  = "CONFLICT_ADVISOR"
    COST_MONITOR      = "COST_MONITOR"
    UPSELL_ASSISTANT  = "UPSELL_ASSISTANT"


class PromoAIRecStatus(str, enum.Enum):
    PENDING      = "PENDING"
    ACKNOWLEDGED = "ACKNOWLEDGED"
    ACTIONED     = "ACTIONED"
    DISMISSED    = "DISMISSED"


# ── Promotional Scheme Header ──────────────────────────────────────────────────

class PromoScheme(Base, TimestampMixin):
    __tablename__ = "promo_schemes"

    id                        = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    scheme_code               = Column(String(50), unique=True, nullable=False)
    scheme_name               = Column(String(200), nullable=False)
    scheme_type               = Column(Enum(SchemeType), nullable=False)
    status                    = Column(Enum(SchemeStatus), default=SchemeStatus.DRAFT)
    valid_from                = Column(Date, nullable=False)
    valid_to                  = Column(Date, nullable=False)
    priority_rank             = Column(Integer, default=10)
    stackable                 = Column(Boolean, default=False)
    exclusive                 = Column(Boolean, default=False)
    requires_approval_override = Column(Boolean, default=False)
    cost_center_id            = Column(UUID(as_uuid=True), nullable=True)
    notes                     = Column(Text)
    approved_by_id            = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    created_by_id             = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)

    eligibility_scopes = relationship("PromoEligibility", back_populates="scheme", cascade="all, delete-orphan")
    rule_lines         = relationship("PromoRuleLine", back_populates="scheme", cascade="all, delete-orphan")
    order_applications = relationship("SalesOrderPromo", back_populates="scheme")


# ── Eligibility Scope ──────────────────────────────────────────────────────────

class PromoEligibility(Base, TimestampMixin):
    __tablename__ = "promo_eligibilities"

    id                         = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    scheme_id                  = Column(UUID(as_uuid=True), ForeignKey("promo_schemes.id"), nullable=False)
    applies_to_customer_id     = Column(UUID(as_uuid=True), nullable=True)
    applies_to_customer_group  = Column(String(100), nullable=True)
    applies_to_distributor_group = Column(String(100), nullable=True)
    applies_to_region          = Column(String(100), nullable=True)
    applies_to_channel         = Column(String(100), nullable=True)
    applies_to_sales_team      = Column(String(100), nullable=True)
    applies_to_price_list_id   = Column(UUID(as_uuid=True), nullable=True)
    applies_to_item_id         = Column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=True)
    applies_to_item_category   = Column(String(100), nullable=True)
    applies_to_brand           = Column(String(100), nullable=True)
    min_order_qty              = Column(Numeric(14, 3))
    min_order_value            = Column(Numeric(14, 2))
    active                     = Column(Boolean, default=True)
    notes                      = Column(Text)

    scheme = relationship("PromoScheme", back_populates="eligibility_scopes")


# ── Promotion Rule Lines ───────────────────────────────────────────────────────

class PromoRuleLine(Base, TimestampMixin):
    __tablename__ = "promo_rule_lines"

    id                        = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    scheme_id                 = Column(UUID(as_uuid=True), ForeignKey("promo_schemes.id"), nullable=False)
    trigger_basis             = Column(Enum(TriggerBasis), nullable=False, default=TriggerBasis.SKU)
    trigger_item_id           = Column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=True)
    trigger_category          = Column(String(100))
    trigger_brand             = Column(String(100))
    min_trigger_qty           = Column(Numeric(14, 3), default=0)
    min_trigger_value         = Column(Numeric(14, 2), default=0)
    max_trigger_qty           = Column(Numeric(14, 3))
    reward_type               = Column(Enum(RewardType), nullable=False)
    reward_item_id            = Column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=True)
    reward_qty                = Column(Numeric(14, 3))
    reward_percent            = Column(Numeric(6, 4))
    reward_amount             = Column(Numeric(14, 2))
    reward_special_unit_price = Column(Numeric(14, 4))
    max_reward_qty            = Column(Numeric(14, 3))
    repeatable                = Column(Boolean, default=False)
    sort_order                = Column(Integer, default=1)
    notes                     = Column(Text)

    scheme       = relationship("PromoScheme", back_populates="rule_lines")
    trigger_item = relationship("Product", foreign_keys=[trigger_item_id])
    reward_item  = relationship("Product", foreign_keys=[reward_item_id])


# ── Tiered Discount Lines ──────────────────────────────────────────────────────

class PromoTierLine(Base, TimestampMixin):
    __tablename__ = "promo_tier_lines"

    id              = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    rule_line_id    = Column(UUID(as_uuid=True), ForeignKey("promo_rule_lines.id"), nullable=False)
    min_qty         = Column(Numeric(14, 3), nullable=False)
    max_qty         = Column(Numeric(14, 3))
    min_value       = Column(Numeric(14, 2))
    max_value       = Column(Numeric(14, 2))
    reward_percent  = Column(Numeric(6, 4))
    reward_amount   = Column(Numeric(14, 2))
    unit_price      = Column(Numeric(14, 4))
    sort_order      = Column(Integer, default=1)

    rule_line = relationship("PromoRuleLine")


# ── Sales Order Promo Application ─────────────────────────────────────────────

class SalesOrderPromo(Base, TimestampMixin):
    __tablename__ = "sales_order_promos"

    id                       = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    sales_order_id           = Column(UUID(as_uuid=True), ForeignKey("sales_orders.id"), nullable=False)
    scheme_id                = Column(UUID(as_uuid=True), ForeignKey("promo_schemes.id"), nullable=False)
    application_type         = Column(Enum(PromoApplicationType), default=PromoApplicationType.AUTO)
    calculated_benefit_amount = Column(Numeric(14, 2), default=0)
    promo_cost_estimate      = Column(Numeric(14, 2), default=0)
    stack_sequence           = Column(Integer, default=1)
    applied_by_id            = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    notes                    = Column(Text)

    scheme      = relationship("PromoScheme", back_populates="order_applications")
    line_impacts = relationship("SalesOrderPromoLine", back_populates="order_promo", cascade="all, delete-orphan")


# ── Sales Order Promo Line Impact ─────────────────────────────────────────────

class SalesOrderPromoLine(Base, TimestampMixin):
    __tablename__ = "sales_order_promo_lines"

    id               = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    order_promo_id   = Column(UUID(as_uuid=True), ForeignKey("sales_order_promos.id"), nullable=False)
    sales_order_line_id = Column(UUID(as_uuid=True), nullable=True)
    source_rule_id   = Column(UUID(as_uuid=True), ForeignKey("promo_rule_lines.id"), nullable=True)
    impact_type      = Column(Enum(PromoImpactType), nullable=False)
    impacted_qty     = Column(Numeric(14, 3))
    impacted_amount  = Column(Numeric(14, 2))
    reward_item_id   = Column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=True)
    reward_qty       = Column(Numeric(14, 3))
    notes            = Column(Text)

    order_promo  = relationship("SalesOrderPromo", back_populates="line_impacts")
    reward_item  = relationship("Product", foreign_keys=[reward_item_id])
    source_rule  = relationship("PromoRuleLine", foreign_keys=[source_rule_id])


# ── Promo Override Request ────────────────────────────────────────────────────

class PromoOverrideRequest(Base, TimestampMixin):
    __tablename__ = "promo_override_requests"

    id                  = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    sales_order_id      = Column(UUID(as_uuid=True), ForeignKey("sales_orders.id"), nullable=False)
    scheme_id           = Column(UUID(as_uuid=True), ForeignKey("promo_schemes.id"), nullable=True)
    requested_by_id     = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    approved_by_id      = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    status              = Column(Enum(OverrideStatus), default=OverrideStatus.PENDING)
    requested_discount_pct = Column(Numeric(6, 4))
    requested_free_qty  = Column(Numeric(14, 3))
    reason              = Column(Text, nullable=False)
    approver_notes      = Column(Text)

    scheme = relationship("PromoScheme")


# ── Scheme Usage Tally ────────────────────────────────────────────────────────

class PromoUsageTally(Base, TimestampMixin):
    __tablename__ = "promo_usage_tallies"
    __table_args__ = (UniqueConstraint("scheme_id", "tally_month"),)

    id               = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    scheme_id        = Column(UUID(as_uuid=True), ForeignKey("promo_schemes.id"), nullable=False)
    tally_month      = Column(String(7), nullable=False)  # YYYY-MM
    order_count      = Column(Integer, default=0)
    total_discount   = Column(Numeric(14, 2), default=0)
    total_free_value = Column(Numeric(14, 2), default=0)
    total_cost       = Column(Numeric(14, 2), default=0)

    scheme = relationship("PromoScheme")


# ── AI Recommendations ────────────────────────────────────────────────────────

class PromoAIRecommendation(Base, TimestampMixin):
    __tablename__ = "promo_ai_recommendations"

    id          = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    agent_type  = Column(Enum(PromoAIAgentType), nullable=False)
    title       = Column(String(300), nullable=False)
    detail      = Column(Text)
    severity    = Column(String(20), default="info")
    status      = Column(Enum(PromoAIRecStatus), default=PromoAIRecStatus.PENDING)
    scheme_id   = Column(UUID(as_uuid=True), ForeignKey("promo_schemes.id"), nullable=True)

File: frontend/src/lib/promotions.ts
import axios from "axios";

const API = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000/api/v1";
const r = <T>(p: Promise<{ data: T }>) => p.then((res) => res.data);
const ax = axios.create({ baseURL: API });

// ── Enums & Types ──────────────────────────────────────────────────────────────

export type SchemeStatus = "DRAFT" | "APPROVED" | "ACTIVE" | "EXPIRED" | "SUSPENDED" | "ARCHIVED";
export type SchemeType =
  | "BUY_X_GET_Y" | "PERCENT_DISCOUNT" | "FIXED_DISCOUNT" | "TIERED_DISCOUNT"
  | "QTY_BREAK_PRICE" | "SPEND_BASED" | "MIX_AND_MATCH" | "BUNDLE"
  | "FREE_GOODS_DIFF_SKU" | "CHANNEL_DEAL";
export type TriggerBasis = "SKU" | "CATEGORY" | "BRAND" | "ORDER_TOTAL" | "QUANTITY" | "VALUE" | "MIX_SET";
export type RewardType = "FREE_GOODS" | "PERCENT_DISCOUNT" | "FIXED_DISCOUNT" | "SPECIAL_PRICE" | "BUNDLE_PRICE";
export type PromoApplicationType = "AUTO" | "MANUAL_APPROVED" | "MANUAL_OVERRIDE" | "REJECTED";
export type PromoImpactType = "DISCOUNT" | "FREE_GOODS" | "SPECIAL_PRICE" | "ORDER_DISCOUNT" | "BUNDLE_ADJUSTMENT";
export type OverrideStatus = "PENDING" | "APPROVED" | "REJECTED";
export type PromoAIAgentType = "CONFLICT_ADVISOR" | "COST_MONITOR" | "UPSELL_ASSISTANT";
export type PromoAIRecStatus = "PENDING" | "ACKNOWLEDGED" | "ACTIONED" | "DISMISSED";

// ── Interfaces ─────────────────────────────────────────────────────────────────

export interface PromoTierLine {
  id: string;
  min_qty?: number;
  max_qty?: number;
  min_value?: number;
  max_value?: number;
  reward_percent?: number;
  reward_amount?: number;
  unit_price?: number;
  sort_order: number;
}

export interface PromoRuleLine {
  id: string;
  scheme_id: string;
  trigger_basis: TriggerBasis;
  trigger_item_id?: string;
  trigger_item_name?: string;
  trigger_category?: string;
  trigger_brand?: string;
  min_trigger_qty: number;
  min_trigger_value: number;
  max_trigger_qty?: number;
  reward_type: RewardType;
  reward_item_id?: string;
  reward_item_name?: string;
  reward_qty?: number;
  reward_percent?: number;
  reward_amount?: number;
  reward_special_unit_price?: number;
  max_reward_qty?: number;
  repeatable: boolean;
  sort_order: number;
  notes?: string;
}

export interface PromoEligibility {
  id: string;
  scheme_id: string;
  applies_to_customer_id?: string;
  applies_to_customer_group?: string;
  applies_to_distributor_group?: string;
  applies_to_region?: string;
  applies_to_channel?: string;
  applies_to_item_category?: string;
  applies_to_brand?: string;
  min_order_qty?: number;
  min_order_value?: number;
  active: boolean;
}

export interface PromoScheme {
  id: string;
  scheme_code: string;
  scheme_name: string;
  scheme_type: SchemeType;
  status: SchemeStatus;
  valid_from: string;
  valid_to: string;
  priority_rank: number;
  stackable: boolean;
  exclusive: boolean;
  requires_approval_override: boolean;
  notes?: string;
  eligibility_scopes: PromoEligibility[];
  rule_lines: PromoRuleLine[];
  usage_count: number;
  total_cost: number;
}

export interface PromoLineImpact {
  id: string;
  impact_type: PromoImpactType;
  impacted_qty?: number;
  impacted_amount?: number;
  reward_item_id?: string;
  reward_item_name?: string;
  reward_qty?: number;
  notes?: string;
}

export interface SalesOrderPromo {
  id: string;
  sales_order_id: string;
  scheme_id: string;
  scheme_name?: string;
  scheme_code?: string;
  application_type: PromoApplicationType;
  calculated_benefit_amount: number;
  promo_cost_estimate: number;
  stack_sequence: number;
  line_impacts: PromoLineImpact[];
}

export interface PromoApplicationResult {
  scheme_id: string;
  scheme_code: string;
  scheme_name: string;
  application_type: PromoApplicationType;
  calculated_benefit: number;
  promo_cost_estimate: number;
  stack_sequence: number;
  line_impacts: Array<{
    impact_type: PromoImpactType;
    reward_item_name?: string;
    reward_qty?: number;
    discount_amount?: number;
    discount_pct?: number;
    notes?: string;
  }>;
  next_threshold_hint?: string;
}

export interface EvaluateOrderResult {
  sales_order_id: string;
  applied_promos: PromoApplicationResult[];
  skipped_promos: Array<{ scheme_id: string; scheme_code?: string; reason: string; next_threshold_hint?: string }>;
  total_discount: number;
  total_free_goods_value: number;
  total_promo_cost: number;
}

export interface OverrideRequest {
  id: string;
  sales_order_id: string;
  scheme_id?: string;
  status: OverrideStatus;
  requested_discount_pct?: number;
  requested_free_qty?: number;
  reason: string;
  approver_notes?: string;
  created_at?: string;
}

export interface PromoAIRec {
  id: string;
  agent_type: PromoAIAgentType;
  title: string;
  detail?: string;
  severity: string;
  status: PromoAIRecStatus;
  scheme_id?: string;
}

export interface PromoDashboard {
  active_schemes: number;
  expiring_soon: number;
  total_applications_month: number;
  total_discount_month: number;
  total_free_value_month: number;
  pending_override_requests: number;
  pending_ai_recs: number;
}

// ── Labels ─────────────────────────────────────────────────────────────────────

export const SCHEME_TYPE_LABEL: Record<SchemeType, string> = {
  BUY_X_GET_Y:         "Buy X Get Y Free",
  PERCENT_DISCOUNT:    "% Discount",
  FIXED_DISCOUNT:      "Fixed Discount",
  TIERED_DISCOUNT:     "Tiered Discount",
  QTY_BREAK_PRICE:     "Qty Break Price",
  SPEND_BASED:         "Spend-Based",
  MIX_AND_MATCH:       "Mix & Match",
  BUNDLE:              "Bundle Deal",
  FREE_GOODS_DIFF_SKU: "Free Goods (Different SKU)",
  CHANNEL_DEAL:        "Channel Deal",
};

export const REWARD_TYPE_LABEL: Record<RewardType, string> = {
  FREE_GOODS:       "Free Goods",
  PERCENT_DISCOUNT: "% Discount",
  FIXED_DISCOUNT:   "Fixed Amount Off",
  SPECIAL_PRICE:    "Special Unit Price",
  BUNDLE_PRICE:     "Bundle Price",
};

export const TRIGGER_BASIS_LABEL: Record<TriggerBasis, string> = {
  SKU:         "Specific SKU",
  CATEGORY:    "Item Category",
  BRAND:       "Brand",
  ORDER_TOTAL: "Order Total (spend-based)",
  QUANTITY:    "Any Quantity",
  VALUE:       "Line Value",
  MIX_SET:     "Mix & Match Set",
};

export const STATUS_BADGE: Record<SchemeStatus, string> = {
  DRAFT:     "bg-gray-100 text-gray-600",
  APPROVED:  "bg-blue-100 text-blue-700",
  ACTIVE:    "bg-green-100 text-green-700",
  EXPIRED:   "bg-red-100 text-red-600",
  SUSPENDED: "bg-yellow-100 text-yellow-700",
  ARCHIVED:  "bg-gray-100 text-gray-500",
};

export const AGENT_LABEL: Record<PromoAIAgentType, string> = {
  CONFLICT_ADVISOR: "Conflict Advisor",
  COST_MONITOR:     "Cost Monitor",
  UPSELL_ASSISTANT: "Upsell Assistant",
};

export const SEVERITY_BADGE: Record<string, string> = {
  info:     "bg-blue-100 text-blue-700",
  warning:  "bg-yellow-100 text-yellow-700",
  critical: "bg-red-100 text-red-700",
};

export const fmtCurrency = (n: number) =>
  `KES ${Number(n).toLocaleString("en-KE", { minimumFractionDigits: 2 })}`;

// ── API Client ─────────────────────────────────────────────────────────────────

export const promoApi = {
  // Dashboard
  getDashboard: () => r<PromoDashboard>(ax.get("/promotions/dashboard")),

  // Schemes
  getSchemes: (status?: SchemeStatus, activeOnly?: boolean) =>
    r<PromoScheme[]>(ax.get("/promotions/schemes", { params: { status, active_only: activeOnly } })),
  getScheme: (id: string) => r<PromoScheme>(ax.get(`/promotions/schemes/${id}`)),
  createScheme: (data: object) => r<PromoScheme>(ax.post("/promotions/schemes", data)),
  activateScheme: (id: string) => r<PromoScheme>(ax.post(`/promotions/schemes/${id}/activate`)),
  updateSchemeStatus: (id: string, status: SchemeStatus) =>
    r<PromoScheme>(ax.patch(`/promotions/schemes/${id}/status`, null, { params: { status } })),

  // Evaluation
  evaluateOrder: (data: object) => r<EvaluateOrderResult>(ax.post("/promotions/evaluate-order", data)),
  getOrderPromos: (orderId: string) =>
    r<SalesOrderPromo[]>(ax.get(`/promotions/sales-orders/${orderId}/promotions`)),
  recalculateOrder: (orderId: string, data: object) =>
    r<EvaluateOrderResult>(ax.post(`/promotions/sales-orders/${orderId}/promotions/recalculate`, data)),

  // Override Requests
  getOverrideRequests: (status?: OverrideStatus) =>
    r<OverrideRequest[]>(ax.get("/promotions/override-requests", { params: { status } })),
  createOverrideRequest: (data: object) =>
    r<OverrideRequest>(ax.post("/promotions/override-requests", data)),
  approveOverride: (id: string, data: { approved: boolean; approver_notes?: string }) =>
    r<OverrideRequest>(ax.post(`/promotions/override-requests/${id}/approve`, data)),

  // Reports
  getUsageReport: () => r<any[]>(ax.get("/promotions/reports/usage")),

  // AI
  runAgents: () => r<{ generated: number }>(ax.post("/promotions/ai/run-agents")),
  getAIRecs: (status?: PromoAIRecStatus) =>
    r<PromoAIRec[]>(ax.get("/promotions/ai/recommendations", { params: { status } })),
  ackAIRec: (id: string, data: { status: PromoAIRecStatus }) =>
    r<PromoAIRec>(ax.patch(`/promotions/ai/recommendations/${id}`, data)),
};
